### PR TITLE
feat: LT-1208 - Transhipment

### PR DESCRIPTION
**Pull request by jfaracik**
* You now have to send application sub_type when creating applications
* Added `exclude` query parameter to countries, sites and external locations (to exclude certain countries eg `GB`)
* Cleanup: removed unused return `address` from `create_site`
* Fix: some picklist fields were optional

### CODE REVIEW

File: applications/helpers.py [hunk 1]
```diff
@@ -56,21 +56,23 @@ def get_application_view_serializer(application: BaseApplication):
         )
 
 
-def get_application_create_serializer(application_type):
-    if application_type == CaseTypeSubTypeEnum.STANDARD:
+def get_application_create_serializer(case_type):
+    sub_type = CaseTypeEnum.reference_to_class(case_type).sub_type
```
Comment (andytaylorkainos) [inline applications/helpers.py]: so we're sending siel, oiel, ogel etc. to this endpoint now instead of standard/open?

Comment (jfaracik) [reply applications/helpers.py]: Yep! As we discussed with @efahy

File: applications/tests/tests_adding_sites.py [hunk 2]
```diff
@@ -166,3 +167,21 @@ def test_add_site_to_application_in_editable_status_success(self, editable_statu
...
+            organisation=self.organisation, case_type_id=CaseTypeEnum.SITL.id
+        )
+
+        url = reverse("applications:application_sites", kwargs={"pk": transhipment.id})
+        data = {"sites": [site.id]}
+        response = self.client.post(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(transhipment.application_sites.count(), 1)
+        self.assertEqual(response.json()["errors"]["sites"][0], "YOU CANT SET GB COUNTRY ON TRANSHIPMENT")
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_adding_sites.py]: This error should point to a lite string

File: applications/tests/tests_external_locations.py [hunk 3]
```diff
@@ -181,3 +182,20 @@ def test_add_external_locations_to_application_in_read_only_status_failure(self,
...
+            organisation=self.organisation, case_type_id=CaseTypeEnum.SITL.id
+        )
+
+        url = reverse("applications:application_external_locations", kwargs={"pk": transhipment.id})
+        data = {"external_locations": [external_location.id]}
+        response = self.client.post(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(transhipment.application_sites.count(), 1)
+        self.assertEqual(response.json()["errors"]["external_locations"][0], "YOU CANT SET GB COUNTRY ON TRANSHIPMENT")
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_external_locations.py]: Lite string

File: applications/views/external_locations.py [hunk 4]
```diff
@@ -127,6 +127,12 @@ def post(self, request, application):
                 data={"external_location": new_location.id, "application": application.id,}
             )
 
+            if new_location.country.id == "GB":
+                return JsonResponse(
+                    data={"errors": {"external_locations": ["YOU CANT SET GB COUNTRY ON TRANSHIPMENT"]}},
```
Comment (BenAndrew-Kainos) [inline applications/views/external_locations.py]: lite string

File: applications/views/external_locations.py [hunk 5]
```diff
@@ -127,6 +127,12 @@ def post(self, request, application):
                 data={"external_location": new_location.id, "application": application.id,}
             )
 
+            if new_location.country.id == "GB":
```
Comment (BenAndrew-Kainos) [inline applications/views/external_locations.py]: Please make GB a constant

File: applications/views/sites.py [hunk 6]
```diff
@@ -123,6 +124,13 @@ def post(self, request, application):
         for new_site in new_sites:
             serializer = SiteOnApplicationCreateSerializer(data={"site": new_site.id, "application": application.id})
 
+            if application.case_type.reference == CaseTypeReferenceEnum.SITL:
+                if new_site.address.country.id == "GB":
+                    return JsonResponse(
+                        data={"errors": {"sites": ["YOU CANT SET GB COUNTRY ON TRANSHIPMENT"]}},
```
Comment (BenAndrew-Kainos) [inline applications/views/sites.py]: String

File: static/countries/tests.py [hunk 7]
```diff
@@ -15,3 +15,13 @@ def test_get_countries(self):
...
         self.assertEqual(countries[0]["name"], "Abu Dhabi")
         self.assertEqual(countries[-1]["name"], "Zimbabwe")
+
+    def test_filter_countries_by_exclude(self):
+        response = self.client.get(self.url + "?exclude=GB&exclude=PL")
+        countries = response.json()["countries"]
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(countries[0]["name"], "Abu Dhabi")
+        self.assertEqual(countries[-1]["name"], "Zimbabwe")
```
Comment (BenAndrew-Kainos) [inline static/countries/tests.py]: This seems heavily reliant on the countries list not changing. Could you fetch the first and last country from the list instead

**Comment by jfaracik** [comment]
> It also seems you don't have a model for Transhipment

Yep no model for transhipment applications, identical to regular SIELs but with sites and locations not being in the UK.

File: static/countries/tests.py [hunk 8]
```diff
@@ -1,17 +1,38 @@
...
 class CountriesTests(DataTestClient):
 
     url = reverse("static:countries:countries")
 
+    def setUp(self):
+        super().setUp()
+        self.countries = None
+        with open(COUNTRIES_FILE, newline="") as csvfile:
+            reader = csv.DictReader(csvfile)
+            # Convert CSV to a list of the names of the countries
```
Comment (andytaylorkainos) [inline static/countries/tests.py]: but why?

Cant you just filter the DB?

**Pull request merged**